### ISSUE TITLE
[PW-8506] Wrong Order status when configuration field 'payment_pre_authorized' is set

### ISSUE THREAD

**Issue by thomas-kl1** [Adyen/adyen-magento2#2097]
**Describe the bug**

Why is the method \Adyen\Payment\Helper\Order::setPrePaymentAuthorized called only when we do have a successful authorization there \Adyen\Payment\Helper\Webhook\AuthorisationWebhookHandler::handleSuccessfulAuthorisation ?
Shouldn't we consider authorized payment as processed?

[See Magento Doc Order Flow](https://docs.magento.com/user-guide/sales/order-workflow.html)

> Step 3: Receive Payment
> The status of the order changes to Processing when payment is received or authorized. Depending on the payment method, you might receive notification when the transaction is authorized or processed.

Today there's the following configuration fields that allows us to set the status by event:
- payment_pre_authorized
- payment_authorized

You may have something like this: 

```xml
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Store:etc/config.xsd">
    <default>
        <payment>
            <adyen_abstract>
                <payment_pre_authorized>pending_payment</payment_pre_authorized>
                <payment_authorized>processing</payment_authorized>
                <payment_cancelled>canceled</payment_cancelled>
            </adyen_abstract>
        </payment>
    </default>
</config>
```

So when a payment is pending authorization it is in pending_payment (so it can be cancelled automatically by Magento after a delay). When the payment is authorized we set the processing status.
However, we can see that the field 'payment_pre_authorized' is used when a payment is authorized and not before we got the notification. 'payment_authorized' is used when there's the auto-capture or the capture. So the fields are misleading and not clear. Obviously I tend to think there's an error and that \Adyen\Payment\Helper\Webhook\AuthorisationWebhookHandler::handleSuccessfulAuthorisation should at its beginning set the status from this field: payment_authorized (when $isFullAmountAuthorized is true)

**To Reproduce**
1. Enable manual capture
2. Set the configuration as above
3. Place an order
4. Order is in payment review, then pending payment
5. Order receive the successful AUTHORIZATION event
6. Don't capture the payment, the order is going to be cancelled by Magento (default value is 480 minutes)

**Expected behavior**
After the successful AUTHORIZATION event, the order is in processing state, so it can't be canceled by the Magento cron.

**Magento version**
Any

**Plugin version**
8.x (8.18.2)

**Additional context**

**Comment by RokPopov** [comment]
Hi @thomas-kl1,

Thank you for opening this issue with us. I have opened an internal ticket to investigate the flow and the suggested approach.

We will update this issue accordingly.

Kind regards,
Rok

**Comment by thomas-kl1** [comment]
Hello @RokPopov 

Any updates?

**Comment by dvasconcelos** [comment]
We are facing the same issue. In the configuration, "Order status: payment authorization" corresponds to the "payment_pre_authorized" config which is used after the authorization event. We can only set status associated to the state "pending" and not "processing". So the order stays "pending" after the authorization.

**Comment by meetvora69** [comment]
We have also faced the same issue, our order is only updated till the pending_payment state, and after that order is automatically closed by magento. Could you please provide status processing during payment_pre_authorized? it's a very urgent thing. we have lost our customer confidence.

**Comment by meetvora69** [comment]
If anyone finds out a patch or any solution to this issue share here, it's a major issue and we have lost our business. Adyen this is a major issue using your module kindly quickly work around it and provide a solution. I have created a ticket but no one provide ETA.

**Comment by thomas-kl1** [comment]
@meetvora69 you can create a new status like "authorized_payment" and assign it to the "pending_payment" state. Then in the pre-authorization status field you can select your new status, so Magento won't try to cancel the orders with pending_payment status. 
Adyen must fix this issue I agree as it had a major impact with our clients too.

**Comment by meetvora69** [comment]
Hello @thomas-kl1,

Thanks for the update but actually some of the orders are still pending state even though the order is authorized, so we need to update this order.

**Comment by thomas-kl1** [comment]
@RokPopov any updates there?

**Comment by meetvora69** [comment]
Hello Team,

kindly do update we have face lost of issue in production site.

**Comment by candemiralp** [comment]
Hello all!

Once again, thank you for your patience during our investigation. We used [this](https://business.adobe.com/blog/basics/magento-order-status) Adobe Commerce page as source of truth while making decisions. Let me share the results directly.

1. `payment_pre_authorized` event is being used for all orders which get Authorisation webhook. This event sets the  state to `new` (status Pending) or `pending_payment` (status Pending Payment).
2. Previous event is followed by `payment_authorized` event if the capture mode is set to auto (immediate). This event sets the state to `processing` (status Processing) and creates an invoice. If the capture mode is set to manual, no invoice will be created and according to Adobe Commerce's post, any order doesn't have invoice or shipment should be `pending`. That means `pending_payment` shouldn't be there in the first place.

According to these, setting to `processing` is not an option. Because, invoice should be considered while moving to `processing` not the authorisation according to Adobe.

If you don't want Magento's cron to close your orders, you should set `Order Status: payment authorisation` config field to `Pending` (please check the attached image). To make config fields' events more clear, I am explaining each title vs event below.

1. Order Status: payment authorisation - `payment_pre_authorized`
2. Order status: payment confirmed - `payment_authorized`

<img width="964" alt="Screenshot 2023-09-27 at 16 04 58" src="https://github.com/Adyen/adyen-magento2/assets/20255503/a811000a-50b1-412e-abe2-0165173b018c">

To conclude, no further action will be taken for version 8 in order not to break anyone's implementation. Could you please kindly verify us whether if the proposed configuration change solves your issue or not?

*** 
**Changes planned later on V9**
We are going to use the result of the investigation on V9. Following changes are on our roadmap of V9 but not the initial version. Please feel free to share your ideas.

1. `pending_payment` state will only be used if the shopper has been redirected to PSP page. When the shopper comes back, related state will be set depending on the authorisation of the payment (invoicing will be related to capture mode still). If the shopper doesn't come back, order state will stay as `pending_payment`, meaning that the cart has been abandoned and Magento's cron can safely cancel the order after X time.
2. If the capture mode is set to auto and no giftcard is involved, synchronous payment response will authorise the order. No need to wait for the Authorisation webhook.
3. Authorised but not captured orders can only have `pending` state (and associated statuses).

Looking forward to hearing your opinions!

Best Regards,
Can

**Comment by candemiralp** [comment]
Hello @thomas-kl1,

Thank you for pointing out that idea. Let me discuss this with the team and I will get back to you soon. If we decide to implement your proposal, it will be a breaking change most probably. We might prefer to add it to V9.

Best Regards,
Can

**Comment by candemiralp** [comment]
Hello @thomas-kl1,

We have implemented the solution we discussed above to V9 of the plugin with #2263. However, that implementation broke partial capture feature of the plugin. Since an invoice has been created with `pending` state after processing authorisation webhook, it is not possible to create further invoices to capture. Because, the full amount has been invoiced in advance.

We are going to revert that implementation and apply another approach. Plugin creates a new status `adyen_authorized` assigned to state `new`. After processing authorisation webhook, this status will be assigned to the order and no invoice will be created. That will help merchants to differentiate between orders with no payments and authorized orders. However, order state and status can not be `processing` at the moment, because there is no invoice. 

After creating an invoice to make a capture request, order state and status will be `payment_review` until getting a `CAPTURE` webhook. This webhook will update the status depending on the capture result.

Coming back to the original issue, `pending_payment` state will not be used for that purpose anymore.

How that solution does sound to you?

Best Regards,
Can

**Comment by thomas-kl1** [comment]
Why don't you just open the invoice without invoicing the amount? The invoice must be paid with the capture.

**Comment by candemiralp** [comment]
Hello @thomas-kl1,

We tried creating an open invoice and paying that invoice after processing capture webhook.

However, as I mentioned on my previous comment, that implementation broke the partial capture (partial invoicing) feature. Each invoice should be created while capturing the items. If you create an open invoice at the beginning, you can't partially invoice items to partially capture because all the amount has been invoiced at that moment. Open or paid invoice doesn't change the situation.

Due to this blocker, creating an invoice is not an option.

Best Regards,
Can

**Comment by candemiralp** [comment]
Hello All!

After a long discussion both internal and external, a solution was implemented to V9 with PRs #2263 and #2316. From now on, `pending_payment` state for `payment_pre_authorized - Payment authorisation` event can not be used. A new status was implemented with this PR and can be used with `new` state to distinguish the authorized but not captured orders - `adyen_authorized - Authorized`.

For V8, unfortunately, this will be breaking change and can't be implemented. Our advise for V8 is using `Pending` status for `Order status: payment authorisation` configuration field. Alternatively, a custom status can be created and attached to `new` state and can be assigned to this configuration field.

Best Regards,
Can
